### PR TITLE
ruri: new, 3.8

### PR DESCRIPTION
**Pull request by dabao1955**
Topic Description
-----------------
* ruri: new, 3.8

Package(s) Affected
-------------------
* ruri: 3.8

Security Update?
----------------

No

Test Build(s) Done
------------------

**Primary Architectures**

- [x] AMD64 `amd64`
- [x] AArch64 `arm64`
- [x] LoongArch 64-bit `loongarch64`

**Secondary Architectures**

Architectural progress for secondary ports does not impede on merging of this topic.

- [x] Loongson 3 `loongson3`
- [x] PowerPC 64-bit (Little Endian) `ppc64el`
- [x] RISC-V 64-bit `riscv64`

### CODE REVIEW

File: app-admin/ruri/autobuild/defines [hunk 1]
```diff
@@ -0,0 +1,6 @@
+PKGNAME=ruri
+PKGSEC=admin
+PKGDEP="libcap libseccomp"
+PKGDES="A light linux container with zero runtime dependency"
```
Comment (MingcongBai) [inline app-admin/ruri/autobuild/defines]: Please be more descriptive and less commercial-like:

A self-contained Linux container implementation

File: app-admin/ruri/autobuild/defines [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+
+ABSTRIP=0
```
Comment (MingcongBai) [inline app-admin/ruri/autobuild/defines]: Why?

File: app-admin/ruri/autobuild/defines [hunk 3]
```diff
@@ -0,0 +1,6 @@
+PKGNAME=ruri
+PKGSEC=admin
+PKGDEP="libcap libseccomp"
```
Comment (MingcongBai) [inline app-admin/ruri/autobuild/defines]: "Zero dependency" but not really?

File: app-admin/ruri/spec [hunk 4]
```diff
@@ -0,0 +1,4 @@
+VER=3.8
+SRCS="git::commit=tags/v$VER;copy-repo=true::https://github.com/Moe-Hacker/ruri"
+CHKSUMS="SKIP"
+REL=1
```
Comment (MingcongBai) [inline app-admin/ruri/spec]: We do not start from REL=1, drop this line for first-time introduction/updates.

Comment (MingcongBai) [reply app-admin/ruri/spec]: Also, missing CHKUPDATE - see [documentation](https://github.com/AOSC-Dev/aosc-findupdate/blob/master/docs/config.zh-CN.md) (I assume you read Chinese as that was what you used in your last bug report). Anitya is preferred.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

File: app-admin/ruri/autobuild/defines (hunk 1 above)
Comment (dabao1955) [reply app-admin/ruri/autobuild/defines]: Ok, since this description comes from the github repository description, I will fix it as soon as possible.

Comment (MingcongBai) [reply app-admin/ruri/autobuild/defines]: I would recommend referencing the [Styling Manual](https://wiki.aosc.io/zh/developer/packaging/package-styling-manual/#ruan-jian-bao-shuo-ming), we have very specific guidelines for most "aesthetic" considerations when it comes to package scripting.

File: app-admin/ruri/autobuild/defines (hunk 3 above)
Comment (dabao1955) [reply app-admin/ruri/autobuild/defines]: https://wiki.aosc.io/zh/developer/packaging/package-styling-manual/#gou-jian-shi-yi-lai

When built as a static binary, no dependencies are required.

File: app-admin/ruri/autobuild/defines (hunk 2 above)
Comment (dabao1955) [reply app-admin/ruri/autobuild/defines]: https://wiki.aosc.io/zh/developer/packaging/package-styling-manual/#er-jin-zhi-bao

File: app-admin/ruri/spec (hunk 4 above)
Comment (dabao1955) [reply app-admin/ruri/spec]: > 另外，缺少 CHKUPDATE - 请参阅[文档](https://github.com/AOSC-Dev/aosc-findupdate/blob/master/docs/config.zh-CN.md)（我假设您阅读中文，因为这是您在上次错误报告中使用的）。Anitya 是首选。



> Also, missing CHKUPDATE - see [documentation](https://github.com/AOSC-Dev/aosc-findupdate/blob/master/docs/config.zh-CN.md) (I assume you read Chinese as that was what you used in your last bug report). Anitya is preferred.

Obviously this isn't the one I'm looking for.
https://release-monitoring.org/project/82162/

So I tried using the github api.

Comment (MingcongBai) [reply app-admin/ruri/spec]: > > 另外，缺少 CHKUPDATE - 请参阅[文档](https://github.com/AOSC-Dev/aosc-findupdate/blob/master/docs/config.zh-CN.md)（我假设您阅读中文，因为这是您在上次错误报告中使用的）。Anitya 是首选。
> 
> > Also, missing CHKUPDATE - see [documentation](https://github.com/AOSC-Dev/aosc-findupdate/blob/master/docs/config.zh-CN.md) (I assume you read Chinese as that was what you used in your last bug report). Anitya is preferred.
> 
> Obviously this isn't the one I'm looking for. https://release-monitoring.org/project/82162/
> 
> So I tried using the github api.

Just create an account and add your own - the GitHub API would require a personal access token/API key to use with `aosc-findupdate` and is undesirable.

File: app-admin/ruri/autobuild/defines (hunk 3 above)
Comment (MingcongBai) [reply app-admin/ruri/autobuild/defines]: > https://wiki.aosc.io/zh/developer/packaging/package-styling-manual/#gou-jian-shi-yi-lai
> 
> When built as a static binary, no dependencies are required.

Got it.

File: app-admin/ruri/autobuild/defines (hunk 2 above)
Comment (MingcongBai) [reply app-admin/ruri/autobuild/defines]: > https://wiki.aosc.io/zh/developer/packaging/package-styling-manual/#er-jin-zhi-bao

You are building from source.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

File: app-admin/ruri/autobuild/defines (hunk 2 above)
Comment (dabao1955) [reply app-admin/ruri/autobuild/defines]: > > https://wiki.aosc.io/zh/developer/packaging/package-styling-manual/#er-jin-zhi-bao
> 
> You are building from source.

So should I remove this line now?

Comment (MingcongBai) [reply app-admin/ruri/autobuild/defines]: Quite obviously, yes.

**Comment by dabao1955** [comment]
```bash
retro_drop_docs
[INFO]: Build template selected: autotools
[INFO]: autotools > Running configure step ...
[INFO]: Copying replacement autotools base files ...
[INFO]: Re-generating Autotools scripts ...
In file included from /usr/bin/autobuild:13
from /usr/lib/autobuild4/proc/50-build-exec.sh:29
from /usr/lib/autobuild4/templates/10-autotools.sh:51
/usr/lib/autobuild4/templates/10-autotools.sh:36: In function `build_autotools_regenerate':
/usr/lib/autobuild4/templates/10-autotools.sh:36: error: command exited with 1
```
Should I apply a patch to autoconf? Or just run configure directly?

**Comment by Cyanoxygen** [comment]
> Should I apply a patch to autoconf?

No you can't, since the build system used in the project does not look like the GNU Autotools suite. Our `autotools` template only supports the GNU Autotools suite, it can not handle hand-written and non-standard `./configure` script.

You need to write a build script manually to `autobuild/build`, despite the project contains a `./configure` file -- it is not generated by autoconf anyway.

**Comment by dabao1955** [comment]
> > Should I apply a patch to autoconf?
> 
> No you can't, since the build system used in the project does not look like the GNU Autotools suite. Our `autotools` template only supports the GNU Autotools suite, it can not handle hand-written and non-standard `./configure` script.
> 
> You need to write a build script manually to `autobuild/build`, despite the project contains a `./configure` file -- it is not generated by autoconf anyway.
> 
> 

Well, I mean to make a patch to switch the build system to autoconf or cmake. I have tested it and it works fine.

File: app-admin/ruri/autobuild/prepare [hunk 5]
```diff
@@ -0,0 +1,2 @@
+abinfo "Configuring..."
```
Comment (MingcongBai) [inline app-admin/ruri/autobuild/prepare]: No. This is now how you are supposed to call configure - if this is not a standard autotools, then don't advertise it as such.

Or, as you suggested, make a patch to make it so.

File: app-admin/ruri/spec [hunk 6]
```diff
@@ -0,0 +1,4 @@
+VER=3.8
+SRCS="git::commit=tags/v$VER;copy-repo=true::https://github.com/Moe-Hacker/ruri"
+CHKSUMS="SKIP"
+CHKUPDATE="github::repo=Moe-Hacker/ruri"
```
Comment (MingcongBai) [inline app-admin/ruri/spec]: Please use anitya.

**Review by MingcongBai — CHANGES_REQUESTED** [review]

File: app-admin/ruri/autobuild/prepare (hunk 5 above)
Comment (dabao1955) [reply app-admin/ruri/autobuild/prepare]: OK.

File: app-admin/ruri/spec (hunk 6 above)
Comment (dabao1955) [reply app-admin/ruri/spec]: Unable to retrieve the latest uersion from upstream!ERROR:"Error when checking for new version:Rate limit was reached.Will be reset in\"2024-12-31 T03:01:53+00:00\".

Comment (MingcongBai) [reply app-admin/ruri/spec]: > Unable to retrieve the latest uersion from upstream!ERROR:"Error when checking for new version:Rate limit was reached.Will be reset in"2024-12-31 T03:01:53+00:00".

Let them be. That ID would update tomorrow, they seem to be having traffic problems.

File: app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch
Comment (MingcongBai) [inline app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch:1]: Will this be upstreamed at some point? If not, this kind of changes might be trouble in the future.

Comment (dabao1955) [reply app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch:1]: I will try it.

Comment (dabao1955) [reply app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch:1]: ~~But I think the possibility of updating the build system is small.~~

File: app-admin/ruri/spec (hunk 6 above)
Comment (dabao1955) [reply app-admin/ruri/spec]: OK.

**Comment by dabao1955** [comment]
However, ruri is open source under the MIT protocol, which means I can switch to the cmake and autoconf downstream that I maintain myself, but this is still considered as a last resort.

File: app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch
Comment (MingcongBai) [reply app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch:1]: > ~But I think the possibility of updating the build system is small.~

Then I would just write a `build` script. If you have no intention in upstreaming this, then I would consider this patch a problem.

**Comment by MingcongBai** [comment]
> However, ruri is open source under the MIT protocol, which means I can switch to the cmake and autoconf downstream that I maintain myself, but this is still considered as a last resort.

Look, we are a distribution and our job is to ship software consistently. Your argument sounds personally motivated - please knock it off.

**Comment by dabao1955** [comment]
> > However, ruri is open source under the MIT protocol, which means I can switch to the cmake and autoconf downstream that I maintain myself, but this is still considered as a last resort.
> 
> Look, we are a distribution and our job is to ship software consistently. Your argument sounds personally motivated - please knock it off.

Okay, I'm sorry I said that.

File: app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch
Comment (dabao1955) [reply app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch:1]: > > ~But I think the possibility of updating the build system is small.~
> 
> Then I would just write a `build` script. If you have no intention in upstreaming this, then I would consider this patch a problem.

The good news is that upstream can build with cmake, but the bad news is that it is not included in the 3.8 tag.

Comment (MingcongBai) [reply app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch:1]: > > > ~But I think the possibility of updating the build system is small.~
> > 
> > 
> > Then I would just write a `build` script. If you have no intention in upstreaming this, then I would consider this patch a problem.
> 
> The good news is that upstream can build with cmake, but the bad news is that it is not included in the 3.8 tag.

I would recommend backporting, if you find it necessary.

Comment (dabao1955) [reply app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch:1]: > I would recommend backporting, if you find it necessary.

So?

Comment (stdmnpkg) [reply app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch:1]: > ~But I think the possibility of updating the build system is small.~

由于现在上游已经合并了你提供的 CMakeLists.txt ，现在应当考虑把上游主分支中的 CMakeLists.txt 作为 patch 引入

Comment (dabao1955) [reply app-admin/ruri/autobuild/patches/0001-ruri-switch-to-CMake.patch:1]: 问题是我才删啊

**Review by stdmnpkg — CHANGES_REQUESTED** [review]
可以在 `defines` 里面用 `ABTYPE=cmake` 指定使用 CMake 模板

**Comment by dabao1955** [comment]
> 可以在 `defines` 里面用 `ABTYPE=cmake` 指定使用 CMake 模板

已修复

~~总感觉autobuild的检测逻辑有点问题，autoconf不应该先检测configure.ac或者Makefile.am啥的吗~~

File: app-admin/ruri/spec (hunk 6 above)
Comment (MingcongBai) [reply app-admin/ruri/spec]: > OK.

So please change it?

Comment (dabao1955) [reply app-admin/ruri/spec:4]: I don't know why break this.

Comment (dabao1955) [reply app-admin/ruri/spec]: > > OK.
> 
> So please change it?

Yes

File: app-admin/ruri/spec [hunk 7]
```diff
@@ -0,0 +1,4 @@
+VER=3.8
+SRCS="git::commit=tags/v$VER;copy-repo=true::https://github.com/Moe-Hacker/ruri"
+CHKSUMS="SKIP"
+CHKUPDATE="anitya::id=376165"
```
Comment (stdmnpkg) [inline app-admin/ruri/spec]: 文件尾应有换行

**Review by stdmnpkg — CHANGES_REQUESTED** [review]

**Comment by stdmnpkg** [comment]
works on my x86-64 machine
```
studentmain@MeowPC6 [ ~ ] $ ruri -r Documents/ciel/.ciel/container/dist/ ls
Check if uidmap is installed on your host, command like su will run failed without uidmap.
bin  boot  debs  dev  etc  home  lib  lib64  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  tree  usr  var
```

File: app-admin/ruri/spec (hunk 7 above)
Comment (dabao1955) [reply app-admin/ruri/spec]: ok

**Review by MingcongBai — APPROVED** [review]

**Review by MingcongBai — CHANGES_REQUESTED** [review]
Code looks good now, please clean up your commit msg in accordance with the [Styling Manual](https://wiki.aosc.io/zh/developer/packaging/package-styling-manual/).

**Review by stdmnpkg — DISMISSED** [review]
works on my loongarch64 machine

```
studentmain@a-long-loooooong-ago [ ~ ] ! ruri -r Documents/ciel/.ciel/container/dist/ ls
Check if uidmap is installed on your host, command like su will run failed without uidmap.
bin  boot  debs  dev  etc  home  lib  lib64  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  tree  usr  var
```

**Comment by dabao1955** [comment]
> Code looks good now, please clean up your commit msg in accordance with the [Styling Manual](https://wiki.aosc.io/zh/developer/packaging/package-styling-manual/).

Could I change pr branch to clean commit?

**Comment by MingcongBai** [comment]
> > Code looks good now, please clean up your commit msg in accordance with the [Styling Manual](https://wiki.aosc.io/zh/developer/packaging/package-styling-manual/).
> 
> Could I change pr branch to clean commit?

No, squash and force push will suffice.

**Comment by dabao1955** [comment]
what

**Comment by Cyanoxygen** [comment]
> > 可以在 `defines` 里面用 `ABTYPE=cmake` 指定使用 CMake 模板
> 
> 已修复
> 
> ~总感觉autobuild的检测逻辑有点问题，autoconf不应该先检测configure.ac或者Makefile.am啥的吗~



> what

For a quick reference:

```sh
git rebase -i stable
```

Then change all "pick"s to "fixup"s for all but the first entry:

```gitrebase
pick commit1 ...
pick commit2 ...
pick commit3 ...
```

```gitrebase
pick commit1 ...
fixup commit2 ...
fixup commit3 ...
```
After that, change the "pick" to "edit" for the first entry:

```
edit commit1 ...
fixup commit2 ...
fixup commit3 ...
```

Then save and exit the editor. You will be prompted to edit the first commit using `git commit --amend`, that's the chance for you to edit the commit message. 

After that, just run `git rebase --continue`, then `git push -f` (you have to force push since the commit history is completely different).

**Review by MingcongBai — APPROVED** [review]

**Pull request merged**